MsixCore/msixmgr/CIMProvider.cpp

```diff
@@ -0,0 +1,128 @@
+#include "MSIXWindows.hpp"
+#include "CIMProvider.hpp"
+#include "ApplyACLsProvider.hpp"
+#include "msixmgrLogger.hpp"
+#include "..\msixmgrLib\GeneralUtil.hpp"
+#include <string>
+#include <iostream>
+#include <vector>
+#include <TraceLoggingProvider.h>
+#include "InstallUI.hpp"
+
+using namespace MsixCoreLib;
+using namespace std;
+
+namespace MsixCoreLib
+{
+    HRESULT CreateAndAddToCIM(
+        _In_ std::wstring cimPath,
+        _In_ std::wstring sourcePath,
+        _In_ std::wstring rootDirectory)
+    {
+        auto autoFreeLibrary = [](HMODULE* module)
+        {
+            if (module != nullptr)
+            {
+                FreeLibrary(*module);
+            }
+        };
+
+        std::unique_ptr<HMODULE, decltype(autoFreeLibrary)>
+            createCIMDll(nullptr, autoFreeLibrary);
+
+        HMODULE createCIMLocal = LoadLibrary(L"createcim.dll");
+        createCIMDll.reset(&createCIMLocal);
+
+        if (*createCIMDll == nullptr)
+        {
+            std::wcout << "Failed to load createcim.dll. Please confirm the dll is next to this exe." << std::endl;
+            return HRESULT_FROM_WIN32(ERROR_MOD_NOT_FOUND);
+        }
+
+        typedef HRESULT(STDMETHODCALLTYPE *CREATEANDADDTOCIMFILE)(
+            std::wstring cimFilePath,
+            std::wstring sourceRootPath,
+            std::wstring imageRootPath);
+
+        CREATEANDADDTOCIMFILE CreateAndAddToCimFileFunc =
+            reinterpret_cast<CREATEANDADDTOCIMFILE>
+            (GetProcAddress(*createCIMDll, "CreateAndAddToCIMFile"));
+        
+        RETURN_IF_FAILED(CreateAndAddToCimFileFunc(cimPath, sourcePath, rootDirectory));
+
+        return S_OK;
+    }
+
+    HRESULT MountCIM(
+        std::wstring cimFilePath,
+        GUID volumeId)
+    {
+        auto autoFreeLibrary = [](HMODULE* module)
+        {
+            if (module != nullptr)
+            {
+                FreeLibrary(*module);
+            }
+        };
+
+        std::unique_ptr<HMODULE, decltype(autoFreeLibrary)>
+            createCIMDll(nullptr, autoFreeLibrary);
+
+        HMODULE createCIMLocal = LoadLibrary(L"createcim.dll");
+        createCIMDll.reset(&createCIMLocal);
+
+        if (*createCIMDll == nullptr)
+        {
+            std::wcout << "Failed to load createcim.dll. Please confirm the dll is next to this exe." << std::endl;
+            return HRESULT_FROM_WIN32(ERROR_MOD_NOT_FOUND);
+        }

```

Comment:
it seems this is repeated for every function you call, maybe use a class to wrap the loadlibrary and have it free in the destructor?

Reply:
Created a new class to wrap the loadlibrary - let me know if this isn't what you had in mind.

Reply:
yes, that looks better; just because it's in a new class doesn't mean you can't have the output